Implement Gorilla for float compression

aabduvakhobov opened the pull request:
This PR implements the support for compressing float data type by Gorilla compression method.  It uses ELF compression method's implementation of Gorilla for floats: https://github.com/Spatio-Temporal-Lab/elf

Currently, ModelarDB implements Gorilla compression method for doubles, which uses a higher number of bits for storing leading and trailing zeros. This PR ensures that ModelarDB uses only the necessary amount of bits for compressing floats, thus improving its compression.

skejserjensen commented on crates/modelardb_compression/src/models/gorilla.rs:
```diff
@@ -110,8 +112,12 @@ impl Gorilla {
-            let leading_zero_bits = value_xor_last_value.leading_zeros() as u8;
+            let mut leading_zero_bits = value_xor_last_value.leading_zeros() as u8;
             let trailing_zero_bits = value_xor_last_value.trailing_zeros() as u8;
+            // Avoid overflow of leading bits.
```
Maybe add a newline before the comment, extend the comment to say "leading zero bits", and extend the comment to describe why the "leading zero bits" may overflow if this is not done.

skejserjensen commented on crates/modelardb_compression/src/models/gorilla.rs:
```diff
@@ -128,15 +134,20 @@ impl Gorilla {
                     models::VALUE_SIZE_IN_BITS - leading_zero_bits - trailing_zero_bits;
-                self.compressed_values
-                    .append_bits(meaningful_bits as u64, 6);
+                // 0 is used to denote 32 meaningful bits.
```
Maybe add a newline before the comment.

skejserjensen commented on crates/modelardb_compression/src/models/gorilla.rs:
```diff
@@ -128,15 +134,20 @@ impl Gorilla {
-                self.compressed_values
-                    .append_bits(meaningful_bits as u64, 6);
+                // 0 is used to denote 32 meaningful bits.
+                if meaningful_bits == 32 {
```
Is a branch needed? Could it not be computed as `meaningful_bits % 32` to avoid the branch?

skejserjensen reviewed (changes requested):
Thank you very much for the PR. Do you have results showing how much these changes impact ingestion time and compression for real-life data sets? I compared this PR to the current `main` branch and I did not see an improvement to the compression, but ingestion time seemed to increase by around 1.5-3%.

aabduvakhobov commented:
I did some tests to check if ELF's adaptation of Gorilla for float compression (EGF) improves ModelarDB's compression. I did tests using two proprietary datasets and a public dataset. 

Compression results showed EGF compresses slightly worse than ModelarDB's current implementation for lower error bounds like 0% and 0.01% (uses values and bit rewriting).  I've seen a slight improvement in compression for higher error bounds like 1% and 5% when Gorilla is combined with value and bit rewriting. However, with these error bounds, the number of values compressed by Gorilla gets very small due to PMC and Swing efficiency.

As for ingestion speed, I also noticed 1-3% decline with EGF even after addressing the comments in the PR thread. To conclude, I don't think it is worth to replace ModelarDB's current Gorilla implementation with EGF. Thus, I am closing this PR.
